File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
def solution(record):
    answer = []
    leave = []
    name = {}
    
    for re in record:
        re = re.split(" ")
        
        if re[0] == "Enter":
            name[re[1]] = re[2]
            leave += [re[1]+ "님이 들어왔습니다."]
        elif re[0] == "Leave":
            leave += [re[1]+ "님이 나갔습니다."]
        else:
            name[re[1]] = re[2]
    
    for l in leave:
        l = l.replace(l[0:l.index("님")], name[l[0:l.index("님")]])
        answer.append(l)
        
    return answer

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <sstream>

using namespace std;

vector<string> solution(vector<string> record) {
    vector<string> answer;
    unordered_map<string, string> users;
    
    for (int i=0; i<record.size(); i++){
        stringstream ss(record[i]);
        string state, uid, name;
        ss >> state >> uid;

        if (state == "Enter" or state == "Change"){
            ss >> name;
            users[uid] = name;
        }
    }
    
    for (int i=0; i<record.size(); i++){
        stringstream ss(record[i]);
        string state, uid;
        ss >> state >> uid;

        if (state == "Enter"){
            answer.push_back(users[uid] + "님이 들어왔습니다.");
        }
        else if (state == "Leave"){
            answer.push_back(users[uid] + "님이 나갔습니다.");
        }
    }
    return answer;
}